storage/innobase/buf/buf0buf.cc

```diff
@@ -5885,27 +5891,29 @@ static void buf_pool_invalidate_instance(buf_pool_t *buf_pool) {
 
   ut_ad(!mutex_own(&buf_pool->LRU_list_mutex));
 
+  os_event_reset(buf_pool->no_invalidate);
+
+  BOOST_SCOPE_EXIT_ALL(&) { os_event_set(buf_pool->no_invalidate); };
+
   mutex_enter(&buf_pool->flush_state_mutex);
 
   for (i = BUF_FLUSH_LRU; i < BUF_FLUSH_N_TYPES; i++) {
-    /* As this function is called during startup and
-    during redo application phase during recovery, InnoDB
-    is single threaded (apart from IO helper threads) at
-    this stage. No new write batch can be in intialization
-    stage at this point. */
-    ut_ad(buf_pool->init_flush[i] == FALSE);
+    /* Although this function is called during startup and
+    during redo application phase during recovery, Percona InnoDB
+    might be running several LRU manager threads at this stage.
+    Hence, a new write batch can be in initialization stage at this point. */
 
-    /* However, it is possible that a write batch that has
-    been posted earlier is still not complete. For buffer
-    pool invalidation to proceed we must ensure there is NO
+    /* For buffer pool invalidation to proceed we must ensure there is NO
     write activity happening. */
-    if (buf_pool->n_flush[i] > 0) {
+    if (buf_pool->n_flush[i] > 0 || buf_pool->init_flush[i]) {
```

Comment:
we just suppressed the assertion. I don't understand the purpose of the event. What exactly it is helping us with?

if you stopped the LRU thread, then this code shouldn't change?

Reply:
@satya-bodapati, the event is reset to notify the LRU thread to stop the execution. Please, mind the following steps:

- We enter `buf_pool_invalidate_instance`. LRU thread is still running.
- We reset event `no_invalidate` to notify the LRU thread to stop. Mind that:
    - LRU might be still running so we have to remove the assertion `buf_pool->init_flush[i] == FALSE`
- If LRU is still running (we find out be checking `n_flush` and `init_flush` we wait for it to end
- At this point we are sure that the LRU is not running: In the previous step we waited for it to end and after that, the LUR thread is waiting for the event `no_invalidate` to be set.

Reply:
got it @ldonoso in the normal case (not invalidate), how do we ensure the LRU thread doesn't simply wait for flush_event to be set?

Reply:
Hmm, once set, it will never wait unless reset?

Reply:
Yes, that's right.

Please, mind that after the wait is finished the event is *not* reset automatically so the next time the LRU thread calls `os_event_wait` it will find it set unless `buf_pool_invalidate_instance` has reset it.